relay/writer/file_util.go

```diff
@@ -0,0 +1,268 @@
+	}
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "open file %s", filename)
+	}
+	defer f.Close()
+
+	// check whether only the file header
+	fileHeaderLen := len(replication.BinLogFileHeader)
+	fs, err := f.Stat()
+	if err != nil {
+		return false, errors.Errorf("get stat for %s", filename)
+	} else if fs.Size() == int64(fileHeaderLen) {
+		return false, nil // only the file header
+	}
+
+	// seek to the beginning of the FormatDescriptionEvent
+	_, err = f.Seek(int64(fileHeaderLen), io.SeekStart)
+	if err != nil {
+		return false, errors.Annotatef(err, "seek to %d for %s", fileHeaderLen, filename)
+	}
+
+	// parse a FormatDescriptionEvent
+	var found bool
+	onEventFunc := func(e *replication.BinlogEvent) error {
+		if e.Header.EventType != replication.FORMAT_DESCRIPTION_EVENT {
+			return errors.Errorf("got %+v, expect FormatDescriptionEvent", e.Header)
+		}
+		found = true
+		return nil
+	}
+
+	eof, err := replication.NewBinlogParser().ParseSingleEvent(f, onEventFunc)
+	if err != nil {
+		return false, errors.Annotatef(err, "parse %s", filename)
+	} else if eof {
+		return false, errors.Annotatef(io.EOF, "parse %s", filename)
+	}
+	return found, nil
+}
+
+// checkIsDuplicateEvent checks if the event is a duplicate event in the file.
+// NOTE: handle cases when file size > 4GB
+func checkIsDuplicateEvent(filename string, ev *replication.BinlogEvent) (bool, error) {
+	// 1. check event start/end pos with the file size
+	fs, err := os.Stat(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "get stat for %s", filename)
+	}
+	evStartPos := int64(ev.Header.LogPos - ev.Header.EventSize)
+	evEndPos := int64(ev.Header.LogPos)
+	if fs.Size() <= evStartPos {
+		return false, nil // the event not in the file
+	} else if fs.Size() < evEndPos {
+		// the file can not hold the whole event, often because the file is corrupt
+		return false, errors.Errorf("file size %d is between event's start pos (%d) and end pos (%d)",
+			fs.Size(), evStartPos, evEndPos)
+	}
+
+	// 2. use a FileReader to read events from the start pos of the passed-in event
+	rCfg := &reader.FileReaderConfig{
+		EnableRawMode: true,
+	}
+	r := reader.NewFileReader(rCfg)
+	defer r.Close()
+	syncStartPos := gmysql.Position{Name: filename, Pos: uint32(evStartPos)}
+	err = r.StartSyncByPos(syncStartPos)
+	if err != nil {
+		return false, errors.Annotatef(err, "start to reader event from %s", syncStartPos)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second) // read the file should not be too slow.
+	defer cancel()
+	var ev2 *replication.BinlogEvent
+	for {
+		ev2, err = r.GetEvent(ctx)
+		if err != nil {
+			return false, errors.Annotatef(err, "get event from %s", syncStartPos)
+		} else if ev2 == nil {
+			// in fact, this should not happen
+			return false, errors.Errorf("get none event from %s", syncStartPos)
+		}
+		if evStartPos > int64(len(replication.BinLogFileHeader)) &&
```

Comment:
do we need `evStartPos > int64(len(replication.BinLogFileHeader))`?

Reply:
needed, if `evStartPos <= int64(len(replication.BinLogFileHeader))`, then we are check the FORMAT_DESCRIPTION_EVENT itself, so should not skip it.

Reply:
start position of real `FORMAT_DESCRIPTION_EVENT` is `len(replication.BinLogFileHeader)`, right?  and can we use `len(replication.BinLogFileHeader)` instead of `checkFormatDescriptionEventExist`?